### PR TITLE
clamp htmlwidget height; tweak propagation of wheel events

### PR DESCRIPTION
**Pull request by kevinushey**
Closes https://github.com/rstudio/rstudio/issues/2174.

Note that this does tweak the scroll behavior for HTML widgets in Notebooks: widgets no longer unilaterally propagate wheel events; rather, they only propagate them when they have reached their scroll bounds and cannot be scrolled anymore.

(I initially wanted to test this by looking at e.g. the height of the iframe, but that was always reported as 0 -- or perhaps I'm misunderstanding some interplay between how an iframe's scroll position is reported)

### CODE REVIEW

**Comment by jmcphers** [comment]
We used to do almost exactly this, and we turned it off intentionally because the double vertical scrollbar experience is very unpleasant; in addition to the weirdness of having two scrollbars right next to each other, it makes scroll events ambiguous (did you mean to scroll the editor surface, or the chunk contents?), and scrolling kind of fiddly. There's still a `MAX_CHUNK_HEIGHT` constant, though I don't think it's used any more:

https://github.com/rstudio/rstudio/blob/2d36df72dde5beb5326b40dbc4d24d58fa922919/src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/ChunkOutputUi.java#L264

I think @javierluraschi turned this off in https://github.com/rstudio/rstudio/commit/8355ca7221acf6f60fd46910cc5b84ce465b827f#diff-87d1d99a586cdc6353e7d3188d20305b. 

Realize that allowing the output to grow without bounds has its own set of downsides, too, but if we're going to tweak the experience we should try to find a third way (let me know what you think).

**Comment by kevinushey** [comment]
Should we still clamp the widget height, then? Otherwise this can happen (zoomed out for dramatic effect):

screen shot 2018-03-29 at 10 36 02 am

But if we clamp the widget height, then the widget can no longer be scrolled and so will be cut off at the bottom, which will be annoying for users as well.

The logic I added doesn't seem adequate for trackpad scroll events, though. I'm not sure how to properly detect how much the inner iframe has been scrolled.

**Comment by jjallaire** [comment]
I think that's just a badly behaved widget (it's equally possible for any R function to produce interminably long output that requires a lot of scrolling)

**Comment by kevinushey** [comment]
We clamp the height of chunk output for that case (and don't allow scrolling), e.g.

screen shot 2018-03-29 at 10 43 05 am

Perhaps we should do the same for HTML widgets?

**Comment by kevinushey** [comment]
Or I can just file this as a bug upstream for `stringr`.

**Comment by jjallaire** [comment]
That sounds right to me.

**Comment by kevinushey** [comment]
I updated the PR to just enforce a maximum height, and I'll file the associated issue upstream with `stringr` to use their own internal scroller if it's really needed.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkOutputWidget.java [hunk 1]
```diff
@@ -327,6 +327,9 @@ public void syncHeight(final boolean scrollToBottom,
       {
          int contentHeight = root_.getElement().getOffsetHeight() + 19;
          height = Math.max(ChunkOutputUi.MIN_CHUNK_HEIGHT, contentHeight);
+         
+         // clamp height of widgets
+         height = Math.min(height, 650);
```
[the remark this answers is not in the document]
Comment (kevinushey) [reply src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkOutputWidget.java]: Yes, it should! (Sorry, I saw you mentioned that earlier as well)

[the remark this answers is not in the document]
Comment (jmcphers) [reply src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkOutputWidget.java]: Great, thanks!

**Pull request merged**